filebeat/input/filestream/internal/input-logfile/harvester_test.go

```diff
@@ -0,0 +1,286 @@
+	"fmt"
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/x-pack/dockerlogbeat/pipelinemock"
+	"github.com/elastic/go-concert/timed"
+	"github.com/elastic/go-concert/unison"
+)
+
+func TestReaderGroup(t *testing.T) {
+	t.Run("assert new group is empty", func(t *testing.T) {
+		rg := newReaderGroup()
+		assert.Equal(t, 0, len(rg.table))
+	})
+
+	t.Run("assert non existent key can be removed", func(t *testing.T) {
+		rg := newReaderGroup()
+		assert.Equal(t, 0, len(rg.table))
+		rg.remove("no such id")
+		assert.Equal(t, 0, len(rg.table))
+	})
+
+	t.Run("assert inserting existing key returns error", func(t *testing.T) {
+		rg := newReaderGroup()
+		ctx, cf, err := rg.newContext("test-id", context.Background())
+		assert.NotNil(t, ctx)
+		assert.NotNil(t, cf)
+		assert.Nil(t, err)
+		assert.Equal(t, 1, len(rg.table))
+
+		newCtx, newCf, err := rg.newContext("test-id", context.Background())
+		assert.Nil(t, newCtx)
+		assert.Nil(t, newCf)
+		assert.Error(t, err)
+	})
+
+	t.Run("assert new key is added, can be removed and its context is cancelled", func(t *testing.T) {
+		rg := newReaderGroup()
+		ctx, cf, err := rg.newContext("test-id", context.Background())
+		assert.NotNil(t, ctx)
+		assert.NotNil(t, cf)
+		assert.Nil(t, err)
+		assert.Equal(t, 1, len(rg.table))
+
+		assert.Nil(t, ctx.Err())
+		rg.remove("test-id")
+
+		assert.Equal(t, 0, len(rg.table))
+		assert.Error(t, ctx.Err(), context.Canceled)
+
+		newCtx, newCf, err := rg.newContext("test-id", context.Background())
+		assert.NotNil(t, newCtx)
+		assert.NotNil(t, newCf)
+		assert.Nil(t, err)
+		assert.Equal(t, 1, len(rg.table))
+		assert.Nil(t, newCtx.Err())
+	})
+}
+
+func TestDefaultHarvesterGroup(t *testing.T) {
+	source := &testSource{"/path/to/test"}
+
+	t.Run("assert a harvester is started in a goroutine", func(t *testing.T) {
+		mockHarvester := &mockHarvester{}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+
+		goroutinesBeforeRun := runtime.NumGoroutine()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			time.Sleep(1 * time.Millisecond)
+		}
+
+		assert.Equal(t, 1, mockHarvester.runCount)
+
+		// when finished removed from bookeeper
+		_, ok := hg.readers.table[source.Name()]
+		assert.False(t, ok)
+		// stopped source can be stopped
+		assert.Nil(t, hg.StopGroup())
+	})
+
+	t.Run("assert a harvester can be stopped and removed from bookkeeper", func(t *testing.T) {
+		mockHarvester := &mockHarvester{}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+
+		goroutinesBeforeRun := runtime.NumGoroutine()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			// wait until harvester is started
+			if mockHarvester.runCount == 1 {
```

Comment:
In general try to favor black box testing. E.g. a callback in the mockHarvester could increase a counter or update a waitGroup. Then we can do some validation and signal a shutdown (mockHarvester would update some other state/WaitGroup on shutdown).